sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/planning/TestPhysicalOperation.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.planning
+
+import org.junit.Assert
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeReference, Expression, ExprId, IsNotNull}
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LocalRelation, LogicalPlan, Project}
+import org.apache.spark.sql.types.{LongType, StringType, TimestampType}
+
+class TestPhysicalOperation extends SparkFunSuite {
+  private val textField = AttributeReference("text1", StringType, nullable = false)()
+  private val dateField = AttributeReference("date1", TimestampType)()
+  private val longField = AttributeReference("long1", LongType)()
+  private val longAliasField = Alias(longField, "long1")(exprId = ExprId(100))
+  private val longAliasField2 =
+    Alias(longField, "long1")(exprId = ExprId(100), qualifier = Seq("tmp"))
+  private val testRelation = LocalRelation(textField, dateField, longField)
+
+  object InvokePhysicalOperation {
+    def apply(plan: LogicalPlan): Seq[Expression] = plan match {
+      case PhysicalOperation(projects, filters, _) =>
+        filters
+      case _ => Nil
+    }
+  }
+
+  test("verify PhysicalOperation returns correct filters with substituted alias when" +
+    " AttributeReferences are same in alias map and filter condition") {
+    val actualFilters = InvokePhysicalOperation(
+      Filter(IsNotNull(longAliasField.toAttribute),
+        Project(Seq(longAliasField, textField),
+          Filter(IsNotNull(dateField),
+            testRelation))))
+    val expectedFilters = Seq(IsNotNull(dateField), IsNotNull(longAliasField))
+
+    Assert.assertEquals(expectedFilters, actualFilters)
+  }
+
+  test("SPARK-29029 verify PhysicalOperation returns correct filters with substituted alias when" +
+    " AttributeReference's ExprId is same but qualifier is different") {
+    val actualFilters = InvokePhysicalOperation(
+      Filter(IsNotNull(longAliasField2.toAttribute),
+        Project(Seq(longAliasField, textField),
+          Filter(IsNotNull(dateField),
+            testRelation))))
+    val expectedFilters = Seq(IsNotNull(dateField), IsNotNull(longAliasField2))
```

Comment:
Can you also show a full reproducer? How did you face such case when using exposed APIs? Catalyst is a set of private APIs.

Reply:
@HyukjinKwon We have a case where we introduce a Custom logical operator for eg `CustomBarrier` in the spark logical plan merely to block some of the Spark optimization rules. However, as part of custom optimizations in `sessionState.experimentalMethods.extraOptimizations` (executed after all spark's optimization rules) we remove the custom logical operator `CustomBarrier` that's when we end up in this case. On digging further we found that the optimizer rule `PushDownPredicate` that handles this case was not getting triggered because of the `CustomBarrier`. Also `PushDownPredicate` is using `AttributeMap` instead of `Map[Attribute, Expression]` hence we do not see such case there and the alias is substituted properly in that case.

Reply:
@HyukjinKwon I have added test suite `DataSourceStrategyWithCustomBarrierSuite` that reproduces the failing scenario. I had to keep the `CustomBarrier` in `Catalyst` package in order to demonstrate the failing scenario.